### PR TITLE
town portal dialog adjustments

### PR DESCRIPTION
**Pull request by Laserlicht**
- right click show town info
- click on list element centers map on object
- show icons for towns

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Perhaps it should also use same town order as on adventure map?

File: client/windows/GUIClasses.cpp [hunk 1]
```diff
@@ -1724,10 +1740,14 @@ CObjectListWindow::CObjectListWindow(const std::vector<int> & _items, std::share
...
 }
 
-CObjectListWindow::CObjectListWindow(const std::vector<std::string> & _items, std::shared_ptr<CIntObject> titleWidget_, std::string _title, std::string _descr, std::function<void(int)> Callback, size_t initialSelection)
+CObjectListWindow::CObjectListWindow(const std::vector<std::string> & _items, std::shared_ptr<CIntObject> titleWidget_, std::string _title, std::string _descr, std::function<void(int)> Callback, size_t initialSelection, std::vector<std::shared_ptr<IImage>> images)
 	: CWindowObject(PLAYER_COLORED, ImagePath::builtin("TPGATE")),
 	onSelect(Callback),
-	selected(initialSelection)
+	onExit(nullptr),
+	onPopup(nullptr),
+	onClicked(nullptr),
```
Comment (IvanSavenko) [inline client/windows/GUIClasses.cpp:1748]: Redundant
```suggestion
```

File: client/windows/GUIClasses.cpp [hunk 2]
```diff
@@ -1701,17 +1704,30 @@ void CObjectListWindow::CItem::select(bool on)
...
+		parent->onPopup(index);
+}
+
+CObjectListWindow::CObjectListWindow(const std::vector<int> & _items, std::shared_ptr<CIntObject> titleWidget_, std::string _title, std::string _descr, std::function<void(int)> Callback, size_t initialSelection, std::vector<std::shared_ptr<IImage>> images)
 	: CWindowObject(PLAYER_COLORED, ImagePath::builtin("TPGATE")),
 	onSelect(Callback),
-	selected(initialSelection)
+	onExit(nullptr),
+	onPopup(nullptr),
+	onClicked(nullptr),
```
Comment (IvanSavenko) [inline client/windows/GUIClasses.cpp:1728]: Same here

**Comment by Laserlicht** [comment]
> Perhaps it should also use same town order as on adventure map?

Added ordering functionality and use localstate. Works - maybe we should find a way to push localstate order back to server in future (would also allow to save order).

**Comment by IvanSavenko** [comment]
Yes, this is known bug, which is a bit pain to fix due to how our serializer works. Also affects hero paths and sleep status.

We should probably save local state in json and push it to server as new netpack whenever this local state changes and let server save it as part of save game. This should avoid all issue we had with this before.

If you are interested in it - feel free to try to implement it.

**Comment by Laserlicht** [comment]
> If you are interested in it - feel free to try to implement it.

Not a high prio for me. But maybe later. :)

**Pull request merged**